desc/singularities.py

```diff
@@ -0,0 +1,733 @@
+        assert islinspaced(
+            src_zeta, axis=0
+        ), "singular integration requires source nodes be equally spaced in zeta"
+
+        # need eval_grid to be linearly spaced in theta, zeta,
+        # and contain only 1 rho value
+        assert isalmostequal(
+            eval_grid.nodes[:, 0]
+        ), "singular integration requires eval grid on a single surface"
+        assert eval_grid.num_nodes == (
+            eval_grid.num_theta * eval_grid.num_zeta
+        ), "singular integration requires a tensor product grid in theta and zeta"
+        eval_theta = eval_grid.nodes[:, 1].reshape(
+            (eval_grid.num_zeta, eval_grid.num_theta)
+        )
+        eval_zeta = eval_grid.nodes[:, 2].reshape(
+            (eval_grid.num_zeta, eval_grid.num_theta)
+        )
+        assert isalmostequal(
+            eval_theta, axis=0
+        ), "singular integration requires rectangular eval grid in theta and zeta"
+        assert isalmostequal(
+            eval_zeta, axis=1
+        ), "singular integration requires rectangular eval grid in theta and zeta"
+        assert islinspaced(
+            eval_theta, axis=1
+        ), "singular integration requires eval nodes be equally spaced in theta"
+        assert islinspaced(
+            eval_zeta, axis=0
+        ), "singular integration requires eval nodes be equally spaced in zeta"
+
+        self._eval_grid = eval_grid
+        self._src_grid = src_grid
+        self._s = s
+        self._q = q
+
+        r, w, dr, dw = _get_quadrature_nodes(q)
+
+        src_dtheta = src_grid.spacing[:, 1]
+        src_dzeta = src_grid.spacing[:, 2] / src_grid.NFP
+
+        self._h_t = jnp.mean(src_dtheta)
+        self._h_z = jnp.mean(src_dzeta)
+
+        self._st = s / 2 * self._h_t * r * jnp.sin(w)
+        self._sz = s / 2 * self._h_z * r * jnp.cos(w)
+
+    def __call__(self, f, i):
+        """Interpolate data to polar grid points.
+
+        Parameters
+        ----------
+        f : ndarray
+            Data at source grid points to interpolate.
+        i : int
+            Index of polar node.
+
+        Returns
+        -------
+        fi : ndarray
+            Source data interpolated to ith polar node.
+        """
+        shp = f.shape[1:]
+        f = f.reshape(
+            (self._src_grid.num_theta, self._src_grid.num_zeta, -1), order="F"
+        )
+        g = fft_interp2d(
+            f,
+            self._eval_grid.num_theta,
+            self._eval_grid.num_zeta,
+            sx=self._st[i],
+            sy=self._sz[i],
+            dx=self._h_t,
+            dy=self._h_z,
+        )
+        return g.reshape((self._eval_grid.num_nodes, *shp), order="F")
+
+
+class DFTInterpolator(_BIESTInterpolator):
+    """Fourier interpolation matrix required for high order singular integration.
+
+    Parameters
+    ----------
+    eval_grid, src_grid : Grid
+        Evaluation and source points for the integral transform.
+        src_grid should be a LinearGrid
```

Comment:
for `dft` the grid can be stellarator symmetric?

Reply:
yeah the eval grid can be anything, but the source grid should still be linear (for accuracy, not for logic)